My partial solution

paulanthonywilson opened the pull request:
Gone a bit K&R with the styling - sorry.

There's some error handling missing (eg malloc failure) and I'm maybe  MAX_STR_LEN is a bit too paranoid.

Just got as far as creating the struct but I'm hungry.

munckymagik commented:
Everything I hoped to see happening 👍 

Would love to see where it takes you from here.

paulanthonywilson commented:
Ok, so I changed the signature of `str_copy` such that it creates a new struct because I didn't want to make assumptions about what was in the destination struct.

I am unclear about what `str_move` is supposed to achieve and what I've implemented has dangers.

¯\\_(ツ)_/¯

munckymagik commented on src/str_t.c:
```diff
@@ -30,8 +20,53 @@ str_t* str_new_from_raw_parts(const char* ptr, size_t len) {
+  // OTOH if it isn't we don't want to be freeing unallocated memory.
+  p_destination->ptr = p_source->ptr;
+  p_destination->len = p_source->len;
+}
```
What happens if someone does `str_free(p_source)` after calling this?

munckymagik commented on tests/test_str_t.c:
```diff
@@ -47,14 +42,83 @@ void test_creating_new_raw_parts() {
+  // accidentally free an array twice;
+
+  str_t* source = str_new_from_zero_terminated("hi");
+  str_t dest;
```
So far we've been allocating `str_t` types on the heap. But this one is on the stack and after calling `str_move` it carries a pointer to heap memory for the string.

paulanthonywilson replied:
True, but it's fine the way it's being used for the test. 

I'm struggling to guess what the purpose of move could be, tbh. I figure that (as usual) it's up to the client to understand what happens to the memory.

munckymagik commented:
> I am unclear about what str_move is supposed to achieve and what I've implemented has dangers.

You're on the right lines. I wanted to do something here to demonstrate the thought processes required for moving ownership of memory. I think this is referred to as _move semantics_ in Rust/C++.

paulanthonywilson commented:
Tried something else but I don't think I understand the purpose of `str_move` yet.

paulanthonywilson commented:
Also my Hungarian is atrocious. I've never been fond of Hungarian, tbh.

munckymagik commented:
> Tried something else but I don't think I understand the purpose of str_move yet.

Do we really need allocate the `str_t` structs?

Have you used C++ much? If so, do we always have to say `string x = new string(..)` or can we just say `string x`? Why? And who decides which is used, end-coder or stdlib?

munckymagik commented:
> Do we really need allocate the str_t structs?

I meant, do we really need to dynamically allocate the `str_t` structs within our string library?

munckymagik commented:
... and your `move` is doing exactly the right thing. You have understood it.
